modules/dop/providers/cms/cicdcmsservice.go

```diff
@@ -127,13 +127,72 @@ func (s *CICDCmsService) CICDCmsCreateOrUpdate(ctx context.Context, req *CICDCms
 		}))
 	}()
 
+	if req.Batch {
+		err := s.deleteNotNeedKeys(ctx, updateRequest.PipelineSource, updateRequest.Ns, keys)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// TODO Use distributed transaction to solve the problem that there is no new addition after deletion
+	// create or update configs
 	if _, err = s.p.PipelineCms.UpdateCmsNsConfigs(utils.WithInternalClientContext(ctx), updateRequest); err != nil {
 		return false, err
 	}
 
 	return true, nil
 }
 
+func (s *CICDCmsService) deleteNotNeedKeys(ctx context.Context, pipelineSource string, ns string, keys []string) error {
+	// get pre configs
+	newContext := apis.WithInternalClientContext(ctx, "dop")
+	preCMS, err := s.p.PipelineCms.GetCmsNsConfigs(newContext, &cmspb.CmsNsConfigsGetRequest{
+		PipelineSource: pipelineSource,
+		Ns:             ns,
+	})
+	if err != nil {
+		return err
+	}
+
+	// get unnecessary keys
+	var deleteKeys []string
+	for _, v := range preCMS.Data {
+		// file type not support change key, so can not delete not find key
+		if v.Type == "FILE" {
```

Comment:
i think this type is wrong.
do you test it?

Reply:
Changed to dice file type

Reply:
It has been tested before, and there is no problem with wrong type judgment